### PR TITLE
Vlad rossohin/members manage progress

### PR DESCRIPTION
**Pull request by VladRossohin**
Review my requests please, there is some code i don't absolutely shure

### CODE REVIEW

File: DIMS/DIMS.BL/Services/vUserProgressService.cs [hunk 1]
```diff
@@ -45,5 +45,13 @@ public vUserProgressDTO GetVUserProgress(int? id)
             return Mapper.Map<vUserProgress, vUserProgressDTO>(_vUserProgress);
 
         }
+
+        public IEnumerable<vUserProgressDTO> GetVUserProgressesByUserId(int? id)
+        {
+            if (!id.HasValue)
+                throw new ValidationException("The view user progress id value is not set", String.Empty);
+
+            return Mapper.Map<IEnumerable<vUserProgress>, List<vUserProgressDTO>>(Database.vUserProgresses.Find(m => m.UserId == id.Value));
```
Comment (maximsan) [inline DIMS/DIMS.BL/Services/vUserProgressService.cs]: From debugging perspective and further support of code I will advise you to split this 1 line in two separated steps: at first get data and then transform/map it. Short code is cool but not good every time.

File: DIMS/DIMS.BL/Services/vUserTaskService.cs [hunk 2]
```diff
@@ -15,18 +16,29 @@ public class vUserTaskService : IvUserTaskService
     {
 
         private IUnitOfWork Database;
```
Comment (maximsan) [inline DIMS/DIMS.BL/Services/vUserTaskService.cs]: field names are camelCase

File: DIMS/DIMS.BL/Services/vUserProfileService.cs [hunk 3]
```diff
@@ -26,20 +26,33 @@ public void Dispose()
...
+            return Mapper.Map<vUserProfile, vUserProfileDTO>(vUserProfile);
         }
 
-        public ICollection<vUserProfileDTO> GetVUserProfiles()
+        public vUserProfileDTO GetVUserProfileByEmail(string email)
+        {
+            if (email == null)
+                throw new ValidationException("The view user profile email is not set", String.Empty);
+
+            var vUserProfile = Database.vUserProfiles.GetByEmail(email);
```
Comment (maximsan) [inline DIMS/DIMS.BL/Services/vUserProfileService.cs]: I will propose you to make some of our methods also async. For example email retrieving, big data fetching and etc. Most of the methods you will use for registration..

File: DIMS/DIMS.DAL.Data/Repositories/vUserProfileRepository.cs [hunk 4]
```diff
@@ -27,6 +27,11 @@ public vUserProfile Get(int id)
             return _dIMSDBContext.vUserProfiles.Find(id);
         }
 
+        public vUserProfile GetByEmail(string email)
+        {
+            return _dIMSDBContext.vUserProfiles.Where(x => x.Email == email).FirstOrDefault();
```
Comment (maximsan) [inline DIMS/DIMS.DAL.Data/Repositories/vUserProfileRepository.cs]: It's not really important, but I for example do not like short names like x, t, p and etc. I'm pretty sure that it's better to give a variable name according to the sense.

File: DIMS/DIMS.DAL.Identity/Repositories/IdentityUnitOfWork.cs [hunk 5]
```diff
@@ -52,5 +52,10 @@ private void Dispose(bool disposing)
                 this.disposed = true;
             }
         }
+
+        public void Save()
+        {
+            _identityDbContext.SaveChanges();
```
Comment (maximsan) [inline DIMS/DIMS.DAL.Identity/Repositories/IdentityUnitOfWork.cs]: Do we really need a sync save too?

File: DIMS/DIMS.Server/Content/styles.css [hunk 6]
```diff
@@ -16,3 +16,10 @@ textarea {
     max-width: 280px;
 }
 
+/*.modal-dialog {
+    padding: 10px;
+    background-color: rgba(200, 200, 200, 0.6);
+    border-radius: 5px;
+}*/
```
Comment (maximsan) [inline DIMS/DIMS.Server/Content/styles.css]: remove

File: DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+            UserProfilesPageViewModel = new UserProfilePageViewModel();
+           /* UserProfilesPageViewModel = new UserProfilePageViewModel
+            {
+                UserProfileViewModel = new UserProfileViewModel(),
+                UserProfilesListViewModel = new UserProfilesListViewModel(),
+                vUserProfilesListViewModel = new vUserProfilesListViewModel(),
+                vUserProfileViewModel = new vUserProfileViewModel(),
+                vUserProgressesListViewModel = new vUserProgressesListViewModel()
+            };*/
```
Comment (maximsan) [inline DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs]: remove

File: DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+        private UserProfilePageViewModel UserProfilesPageViewModel;
+        public UserProfileController(IUserProfileService userProfileService, IvUserProfileService vuserProfileService, IDirectionService directionService, IvUserProgressService vUserProgressService)
+        {
+            _userProfileService = userProfileService;
+            _vUserProfileService = vuserProfileService;
+            _directionService = directionService;
+            _vUserProgressService = vUserProgressService;
+
+            UserProfilesPageViewModel = new UserProfilePageViewModel();
```
Comment (maximsan) [inline DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs]: Why do we not use DI ?

File: DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+            {
+                vUserProfiles = Mapper.Map<IEnumerable<vUserProfileDTO>, List<vUserProfileViewModel>>(vUserProfileDTOs)
+            };
+
+            var directions = _directionService.GetItems();
+
+            UserProfilesPageViewModel.UserProfilesListViewModel = new UserProfilesListViewModel
+            { UserProfiles = Mapper.Map<IEnumerable<UserProfileDTO>, List<UserProfileViewModel>>(userProfileDTOs) };
+            UserProfilesPageViewModel.vUserProfilesListViewModel = new vUserProfilesListViewModel 
+            { vUserProfiles = Mapper.Map<IEnumerable<vUserProfileDTO>, List<vUserProfileViewModel>>(vUserProfileDTOs) };
```
Comment (maximsan) [inline DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs]: please align your code, it looks a little messy

File: DIMS/DIMS.Server/Views/Shared/_Layout.cshtml [hunk 10]
```diff
@@ -4,7 +4,7 @@
     <meta charset="utf-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
     <title>@ViewBag.Title - My ASP.NET Application</title>
-    <link href="~/Content/Site.css" rel="stylesheet" type="text/css" />
+    @*<link href="~/Content/Site.css" rel="stylesheet" type="text/css" />*@
```
Comment (maximsan) [inline DIMS/DIMS.Server/Views/Shared/_Layout.cshtml]: remove

File: DIMS/DIMS.Server/Views/UserProfile/Create.cshtml [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        <div class="modal-body">
+            @using (Html.BeginForm())
+            {
+                @Html.AntiForgeryToken()
+
+            <div class="form-horizontal">
+                
+                @Html.ValidationSummary(true, "", new { @class = "text-danger" })
+
+                @* Name *@
```
Comment (maximsan) [inline DIMS/DIMS.Server/Views/UserProfile/Create.cshtml]: remove

File: DIMS/DIMS.Server/Views/UserProfile/DeleteById.cshtml [hunk 12]
```diff
@@ -0,0 +1,74 @@
+﻿@model HIMS.Server.Models.Users.vUserProfileViewModel
```
Comment (maximsan) [inline DIMS/DIMS.Server/Views/UserProfile/DeleteById.cshtml]: why do we need two views for user deletion ? I'm pretty sure that one is enough fro us

File: DIMS/DIMS.Server/Views/UserProfile/Edit.cshtml [hunk 13]
```diff
@@ -0,0 +1,154 @@
+﻿@model HIMS.Server.Models.Users.UserProfilePageViewModel
```
Comment (maximsan) [inline DIMS/DIMS.Server/Views/UserProfile/Edit.cshtml]: What is the reason to use here the huge UserProfilePageViewModel if you need just UserProfileViewModel ??

File: DIMS/DIMS.Server/Views/UserProfile/Index.cshtml [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+                $.ajaxSetup({ cache: false });
+                $(".delete-user").click(function (e) {
+
+                    e.preventDefault();
+                    $.get(this.href, function (data) {
+                        $('#delete-user-dialogContent').html(data);
+                        $('#delete-user-modal').modal('show');
+                    });
+                });
+            });
```
Comment (maximsan) [inline DIMS/DIMS.Server/Views/UserProfile/Index.cshtml]: You can make you code DRY here, use 1 function everywhere

File: DIMS/DIMS.Server/Views/UserProfile/Index.cshtml [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+        {
+            @Html.AntiForgeryToken()
+
+            <div class="form-actions no-color">
+                <input type="submit" value="Edit" class="btn btn-default" />
+                @Html.ActionLink("Back to List", "Index")
+            </div>
+        }
+    </div>
+</div>
```
Comment (maximsan) [inline DIMS/DIMS.Server/Views/UserProfile/Index.cshtml]: I guess you can also reduce this repetition and make your code DRY

File: DIMS/DIMS.Server/Views/UserProfile/Progress.cshtml [hunk 16]
```diff
@@ -0,0 +1,70 @@
+﻿@model HIMS.Server.Models.Users.UserProfilePageViewModel
```
Comment (maximsan) [inline DIMS/DIMS.Server/Views/UserProfile/Progress.cshtml]: You don't need the huge model here, just use vUserProgresses. If you want to pass some extra data you should use for this ViewData, ViewBag, TempData. Please use them and lighter models.

File: DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+            return PartialView(vuserProfileDto);
+        }
+
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public ActionResult DeleteByEmail(string email, int id)
+        {
+            try
+            {
+                _userProfileService.DeleteUserProfileByEmail(email);
```
Comment (maximsan) [inline DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs]: it's better to make it async

File: DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+            return RedirectToAction("Index");
+        }
+
+        public ActionResult DeleteByEmail(string email)
+        {
+            if (email == null)
+                return new HttpStatusCodeResult(System.Net.HttpStatusCode.BadRequest);
+
+
+            var vuserProfileDto = Mapper.Map<vUserProfileDTO, vUserProfileViewModel>(_vUserProfileService.GetVUserProfileByEmail(email));
```
Comment (maximsan) [inline DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs]: split in two actions and make it async

File: DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+                );
+
+            return PartialView(UserProfilesPageViewModel);
+        }
+
+
+        public ActionResult DeleteById(int? id)
+        {
+            if (!id.HasValue)
+                return new HttpStatusCodeResult(System.Net.HttpStatusCode.BadRequest);
```
Comment (maximsan) [inline DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs]: add namespace

File: DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+        [ValidateAntiForgeryToken]
+        public ActionResult DeleteByEmail(string email, int id)
+        {
+            try
+            {
+                _userProfileService.DeleteUserProfileByEmail(email);
+            }
+            catch (RetryLimitExceededException)
+            {
+                return RedirectToAction("DeleteByEmail", new { email, saveChangesError = true });
```
Comment (maximsan) [inline DIMS/DIMS.Server/Controllers/Users/UserProfileController.cs]: how are you going to treat the error? any message on UI or just throtolling it ?

File: DIMS/DIMS.DAL.Data/Repositories/vUserProfileRepository.cs (hunk 4 above)
Comment (VladRossohin) [reply DIMS/DIMS.DAL.Data/Repositories/vUserProfileRepository.cs]: Ok, I have fixed it

File: DIMS/DIMS.BL/Services/vUserTaskService.cs (hunk 2 above)
Comment (VladRossohin) [reply DIMS/DIMS.BL/Services/vUserTaskService.cs]: done

**Pull request merged**